src/jrd/replication/Config.cpp

```diff
@@ -99,6 +103,69 @@ namespace
 
 		status->setErrors(sv.value());
 	}
+
+	void parseSyncReplica(const ConfigFile::Parameters& params, SyncReplica& output)
+	{
+		for (const auto& el : params)
+		{
+			string key(el.name.c_str());
+			string value(el.value);
+
+			if (value.isEmpty())
+				continue;
+
+			auto pos = key.rfind('_');
+			if (pos != string::npos)
+			{
+				const string key_source = key.substr(pos + 1);
+
+				if (key_source.equals(KEY_ENV))
+				{
+					fb_utils::readenv(value.c_str(), value);
+					if (value.isEmpty())
+						configError("missing environment variable", output.database, value);
+
+					key = key.substr(0, pos);
+				}
+				else if (key_source.equals(KEY_FILE))
+				{
+					PathName filename = value.c_str();
+					PathUtils::fixupSeparators(filename);
+					if (PathUtils::isRelative(filename))
+						filename = fb_utils::getPrefix(IConfigManager::DIR_CONF, filename.c_str());
+
+					AutoPtr<FILE> file(os_utils::fopen(filename.c_str(), "rt"));
+					if (!file)
+						configError("missing or inaccessible file", output.database, filename.c_str());
+
+					// skip first empty lines
+					value = "";
+					do
+					{
+						if (feof(file))
+							break;
+
+						if (!value.LoadFromFile(file))
+							break;
+
+						value.alltrim(" \t\r");
+					} while (value.isEmpty());
+
+					if (value.isEmpty())
+						configError("empty file", output.database, filename.c_str());
+
+					key = key.substr(0, pos);
+				}
+			}
+
+			if (key == "username")
+				output.username = value.c_str();
+			else if (key == "password")
+				output.password = value.c_str();
+			else
+				configError("unknown parameter", output.database, key);
```

Comment:
Is it correct to read the file in advance, before check of key name ?

Reply:
> I think so. Since the configuration is set only by the administrator, it doesn't matter when to read from the file, before or after the key check, but in this case there is no need for double checking the key.

There will be confusion with error diags when wrong key name points to the non-existing file, for example.
Consider case: `pasword_file = deleted_file_name`

> But can be refactor this into a separate function if need?

Sure.